### PR TITLE
Ensure repr and eq for cluster always works

### LINKED ISSUE #4565: Incompatible equality test of addresses in Distributed 2021.3.0
**What happened**:

After updating from 2021.2.0 to 2021.3.0, our regular tests started to observe this error. 
Any pointers as of what might be a good way to mitigate the issue?

````
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/home/yfeng1/source/nbodykit/build/testenv/lib/python3.8/site-packages/nbodykit/tests/test_distributed.py:32: in test_save
    cat.save(tmpfile, dataset='data')
/home/yfeng1/source/nbodykit/build/testenv/lib/python3.8/site-packages/nbodykit/base/catalog.py:692: in save
    source.store(target, regions=region, lock=False, compute=True)
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/array/core.py:1540: in store
    r = store([self], [target], **kwargs)
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/array/core.py:1041: in store
    result.compute(**kwargs)
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/base.py:283: in compute
    (result,) = compute(self, traverse=False, **kwargs)
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/base.py:553: in compute
    schedule = get_scheduler(
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/base.py:1151: in get_scheduler
    return get_scheduler(scheduler=config.get("scheduler", None))
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/dask/base.py:1141: in get_scheduler
    return get_client().get
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/distributed/worker.py:3200: in get_client
    return worker._get_client(timeout=timeout)
/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/distributed/worker.py:3076: in _get_client
    or client._start_arg == self.scheduler.address
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = LocalCluster(d66622f8, 'inproc://192.168.86.31/163961/9', workers=1, threads=1, memory=16.59 GB)
other = 'inproc://192.168.86.31/163961/1'

    def __eq__(self, other):
>       return self.name == other.name
E       AttributeError: 'str' object has no attribute 'name'

/home/yfeng1/.conda/envs/test38/lib/python3.8/site-packages/distributed/deploy/cluster.py:461: AttributeError
```

**What you expected to happen**:

It is not clear how we could triggered malformed input. We are simply creating a LocalCluster in this case. 
Might be a regression on distributed's side. 

**Minimal Complete Verifiable Example**:

```python
# Put your MCVE code here
```

**Anything else we need to know?**:

**Environment**:

- Dask version:  2021.3.0
- Python version: 3.8
- Operating System: Linux
- Install method (conda, pip, source):
Conda

### PR DESCRIPTION
**Pull request by fjetter**
If the scheduler_comm was None, e.g. because the cluster is not connected, yet, there would be an attribute error during `repr`
Also, the equality should verify types since this otherwise also raises unclear exceptions

Closes #4565

I would consider the spawning of a localcluster within a task as highly unorthodox but this has been reported in #4565 and it appears to work. Anyhow, this triggered a weird edge case in the worker connected to the scheduler address and equality which is fixed in this

### CODE REVIEW

File: distributed/tests/test_client.py [hunk 1]
```diff
@@ -6556,3 +6556,26 @@ async def test_workers_collection_restriction(c, s, a, b):
...
+
+    def f(x):
+        with LocalCluster(
+            n_workers=1, processes=False, dashboard_address=False
+        ) as cluster2:
+            with Client(cluster2) as c1:
+                nonlocal scheduler_addr
+                c2 = get_client()
+                # Sometimes the default is closed and we get None for the
+                # scheduler attribute. Prefer robust test case over rigor
```
Comment (jrbourbeau) [inline distributed/tests/test_client.py:6575]: Thanks for taking the time to write this informative comment

File: distributed/tests/test_client.py [hunk 2]
```diff
@@ -6556,3 +6556,26 @@ async def test_workers_collection_restriction(c, s, a, b):
...
+                if c2.status != "closed":
+                    assert c1.scheduler.address != c2.scheduler.address
+                    assert c2.scheduler.address == scheduler_addr
+                    c2.close()
+
+    await c.gather(c.map(f, range(10)))
```
Comment (jrbourbeau) [inline distributed/tests/test_client.py:6581]: Do we need to run `f` 10 times or would once or twice be sufficient? I'm mostly thinking about test runtime

**Review by jrbourbeau — COMMENTED** [review]
Thanks @fjetter! I ran `test_get_client_functions_spawn_clusters` a few times locally and it occasionally failed (see the traceback below). It looks like in some cases `Client.scheduler` can be `None`

<details>
<summary>Traceback:</summary>

```python
(distributed) ➜  distributed git:(repr_eq_cluster) ✗ pytest -v distributed/tests/test_client.py::test_get_client_functions_spawn_clusters --runslow
======================================================================== test session starts =========================================================================
platform darwin -- Python 3.9.2, pytest-6.2.3, py-1.10.0, pluggy-0.13.1 -- /Users/james/miniforge3/envs/distributed/bin/python
cachedir: .pytest_cache
rootdir: /Users/james/projects/dask/distributed, configfile: setup.cfg
plugins: repeat-0.8.0, anyio-2.2.0, rerunfailures-9.1.1, asyncio-0.12.0, timeout-1.4.2
timeout: 300.0s
timeout method: thread
timeout func_only: False
collected 1 item

distributed/tests/test_client.py::test_get_client_functions_spawn_clusters <- distributed/utils_test.py FAILED                                                 [100%]

============================================================================== FAILURES ==============================================================================
______________________________________________________________ test_get_client_functions_spawn_clusters ______________________________________________________________

    def test_func():
        result = None
        workers = []
        with clean(timeout=active_rpc_timeout, **clean_kwargs) as loop:

            async def coro():
                with dask.config.set(config):
                    s = False
                    for _ in range(60):
                        try:
                            s, ws = await start_cluster(
                                nthreads,
                                scheduler,
                                loop,
                                security=security,
                                Worker=Worker,
                                scheduler_kwargs=scheduler_kwargs,
                                worker_kwargs=worker_kwargs,
                            )
                        except Exception as e:
                            logger.error(
                                "Failed to start gen_cluster: "
                                f"{e.__class__.__name__}: {e}; retrying",
                                exc_info=True,
                            )
                            await asyncio.sleep(1)
                        else:
                            workers[:] = ws
                            args = [s] + workers
                            break
                    if s is False:
                        raise Exception("Could not start cluster")
                    if client:
                        c = await Client(
                            s.address,
                            loop=loop,
                            security=security,
                            asynchronous=True,
                            **client_kwargs,
                        )
                        args = [c] + args
                    try:
                        future = func(*args)
                        if timeout:
                            future = asyncio.wait_for(future, timeout)
                        result = await future
                        if s.validate:
                            s.validate_state()
                    finally:
                        if client and c.status not in ("closing", "closed"):
                            await c._close(fast=s.status == Status.closed)
                        await end_cluster(s, workers)
                        await asyncio.wait_for(cleanup_global_workers(), 1)

                    try:
                        c = await default_client()
                    except ValueError:
                        pass
                    else:
                        await c._close(fast=True)

                    def get_unclosed():
                        return [c for c in Comm._instances if not c.closed()] + [
                            c
                            for c in _global_clients.values()
                            if c.status != "closed"
                        ]

                    try:
                        start = time()
                        while time() < start + 60:
                            gc.collect()
                            if not get_unclosed():
                                break
                            await asyncio.sleep(0.05)
                        else:
                            if allow_unclosed:
                                print(f"Unclosed Comms: {get_unclosed()}")
                            else:
                                raise RuntimeError("Unclosed Comms", get_unclosed())
                    finally:
                        Comm._instances.clear()
                        _global_clients.clear()

                    return result

>           result = loop.run_sync(
                coro, timeout=timeout * 2 if timeout else timeout
            )

distributed/utils_test.py:964:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../miniforge3/envs/distributed/lib/python3.9/site-packages/tornado/ioloop.py:530: in run_sync
    return future_cell[0].result()
distributed/utils_test.py:923: in coro
    result = await future
../../../miniforge3/envs/distributed/lib/python3.9/asyncio/tasks.py:481: in wait_for
    return fut.result()
distributed/tests/test_client.py:6581: in test_get_client_functions_spawn_clusters
    await c.gather(c.map(f, range(10)))
distributed/client.py:1840: in _gather
    raise exception.with_traceback(traceback)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   assert c1.scheduler.address != c2.scheduler.address
E   AttributeError: 'NoneType' object has no attribute 'address'

distributed/tests/test_client.py:6577: AttributeError
------------------------------------------------------------------------ Captured stderr call ------------------------------------------------------------------------
distributed.http.proxy - INFO - To route to workers diagnostics web server please install jupyter-server-proxy: python -m pip install jupyter-server-proxy
distributed.scheduler - INFO - Clear task state
distributed.scheduler - INFO -   Scheduler at:     tcp://127.0.0.1:53394
distributed.scheduler - INFO -   dashboard at:            127.0.0.1:8787
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:53395
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:53395
distributed.worker - INFO -          dashboard at:            127.0.0.1:53396
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:53394
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          1
distributed.worker - INFO -                Memory:                  16.00 GiB
distributed.worker - INFO -       Local Directory: /Users/james/projects/dask/distributed/dask-worker-space/worker-0xbamnqu
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:53397
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:53397
distributed.worker - INFO -          dashboard at:            127.0.0.1:53398
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:53394
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          2
distributed.worker - INFO -                Memory:                  16.00 GiB
distributed.worker - INFO -       Local Directory: /Users/james/projects/dask/distributed/dask-worker-space/worker-2gg3260_
distributed.worker - INFO - -------------------------------------------------
distributed.scheduler - INFO - Register worker <Worker 'tcp://127.0.0.1:53395', name: 0, memory: 0, processing: 0>
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:53395
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Register worker <Worker 'tcp://127.0.0.1:53397', name: 1, memory: 0, processing: 0>
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:53397
distributed.core - INFO - Starting established connection
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:53394
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:53394
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Receive client connection: Client-c50b74de-b1ab-11eb-a7bf-acde48001122
distributed.core - INFO - Starting established connection
distributed.worker - WARNING - Compute Failed
Function:  f
args:      (1)
kwargs:    {}
Exception: AttributeError("'NoneType' object has no attribute 'address'")

tornado.application - ERROR - Exception in callback functools.partial(<bound method IOLoop._discard_future_result of <tornado.platform.asyncio.AsyncIOLoop object at 0x11775f7f0>>, <Task finished name='Task-724' coro=<Client._update_scheduler_info() done, defined at /Users/james/projects/dask/distributed/distributed/client.py:1155> exception=AttributeError("'NoneType' object has no attribute 'identity'")>)
Traceback (most recent call last):
  File "/Users/james/miniforge3/envs/distributed/lib/python3.9/site-packages/tornado/ioloop.py", line 741, in _run_callback
    ret = callback()
  File "/Users/james/miniforge3/envs/distributed/lib/python3.9/site-packages/tornado/ioloop.py", line 765, in _discard_future_result
    future.result()
  File "/Users/james/projects/dask/distributed/distributed/client.py", line 1159, in _update_scheduler_info
    self._scheduler_identity = await self.scheduler.identity()
AttributeError: 'NoneType' object has no attribute 'identity'
```

</details>

File: distributed/tests/test_client.py (hunk 2 above)
Comment (fjetter) [reply distributed/tests/test_client.py:6581]: Yes, test runtime concerned me as well but it wasn't significantly lower with fewer iterations. Since I'll need to take another look at the traceback, I'll investigate this as well

File: distributed/deploy/tests/test_cluster.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
+import pytest
+
+from distributed.deploy.cluster import Cluster
+
+
+@pytest.mark.asyncio
+async def test_eq():
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/deploy/tests/test_cluster.py:7]: Are there any cases where cleanup is harmful? We could just turn it to autouse https://docs.pytest.org/en/6.2.x/fixture.html#autouse-fixtures-fixtures-you-don-t-have-to-request

**Comment by fjetter** [comment]
> I ran test_get_client_functions_spawn_clusters a few times locally and it occasionally failed (see the traceback below). It looks like in some cases Client.scheduler can be None

I ran the test more than 100 times and hit this issue only once :(

**Comment by fjetter** [comment]
Test failure is https://github.com/dask/distributed/issues/4766
